### DISCUSSION TITLE
Order of files and order of tabs

### DISCUSSION THREAD

**Discussion by cepamoi** [tsujan/FeatherPad#696] [Show and tell]
Hello.

When FeatherPad is launched with multiple files, the order of the tabs does not reflect the order of the files passed on the command line. For example:
`% featherpad file1 file2 file3 file4`
I would expect the order of the tabs to be: file1 file2 file3 file4
However, the order of the tabs seems to be more or less random.

FeatherPad creates one thread per file to load, and the resulting order of the tabs reflects the order of the end of the execution of the threads, which is unpredictable.

Attached is an **attempt** to solve this problem and to keep the order of the files in the order of the tabs:
[load_order_in_tabs.tar.gz](https://github.com/tsujan/FeatherPad/files/10309446/load_order_in_tabs.tar.gz)
The general idea is to create an object with the ordered list of files while loading the files and use this list to open each new tab at the right position.

In FPsingleton::newWin(), a LoadMultipleFilesOperation object is created. This is basically an ordered list of LoadFileOperation objects. Each LoadFileOperation object will contain a pointer to the corresponding TabPage. A pointer to each LoadFileOperation object is passed through the loading of each file until the FPwin::addText() function. When the loading of a file is completed, the tabs are put in the right order using the LoadMultipleFilesOperation object.

The LoadMultipleFilesOperation is supposed to be deleted when all the files are loaded. This is done in FPwin::addText() with a call to LoadMultipleFilesOperation::destroyIfCompleted(). However, the loading process is complex and there are many possible cases of errors. In such situation, I'm not sure if FPwin::addText() is called and thus if the LoadMultipleFilesOperation is deleted. That's why I added a garbage collector mechanism (see class LoadMultipleFilesOperationGarbageCollector): in worst case, the LoadMultipleFilesOperation objects are deleted at exit.

This tab ordering mechanism is used when loading files from the command line, from the file dialog and when restoring the files from a previous session. There are probably other cases where it could be used, for example with the drag and drop of files from another application.

There might be some issues with the order of the tabs in some (rare) cases. For example when several files are loaded while other files are still being loaded (and the operation is not completed).

And I did not try with Qt 6.

Please review the attached code and let me know what you think.

Thanks.

**Comment by tsujan** [comment]
There's no problem to solve. Please read https://github.com/tsujan/FeatherPad/issues/503#issuecomment-813046548

In short: Use the side-pane mode if the order is important to you.

**Comment by cepamoi** [comment]
The side pane does not solve the problem of the order because:
- The order in the side pane is alphabetical, not related to the order of the files passed to the command line.
- I may prefer tabs instead of the side pane.

The attached code sorts the tabs according to the order of the files passed on the command line.
Would you please consider to have a look at it?
Thanks.

**Reply by tsujan** [reply]
Sorting tabs is possible by using reliable methods that don't introduce bugs/regressions, but I don't add codes for things that have no practical value. Tabs have no predetermined order — they can be dragged and dropped — view-items do.

FeatherPad is open-source. You could always fork it and experiment with your fork.

**Comment by cepamoi** [comment]
It seems that the tab order is important to multiple users as several requests have been made for such a feature: #267, #503 and this discussion. So I would not say that this has no practical value.
Maybe my code is not the right way to implement it.
But you may consider this as a feature request, and implement it in a future version?
Thanks.

**Reply by tsujan** [reply]
> It seems that the tab order is important to multiple users... So I would not say that this has no practical value.

Yes, I know that some users like it to be so, and I respect their preferences — yours too, of course. But there's a difference between preference/habit and practicality. In all years that I've used FeatherPad for coding and other jobs, I haven't seen any practical value to this property. When I open more than 10 files simultaneously and want to find one, I switch to the side-pane mode temporarily.

Tab sorting can be done cleanly by inserting the next tab before or after an existing one appropriately, but it isn't worth the extra codes and computations, especially with a great number of files.

**Reply by cepamoi** [reply]
I'd like to switch to FeatherPad, but keeping tab order is really a missing feature that keeps me from switching.
In my case, I typically work on small examples/tests written in 3 different languages (sometime 4). So I group the files in my text editor by language. And thus I open the files in this way:
`featherpad script1.sh script2.sh file1.cpp file2.cpp file3.cpp perl_file1.pl perl_file2.pl perl_file3.pl`
Here, this is very useful to keep the order from the command line (and the side-pane would not help).
So keeping the file order in the tabs may not have practical value in your case, but it clearly does have some value in mine (and probably for many other users too).

So I would suggest to consider this feature request. You may implement it with an option in the preferences, as already suggested here #[503](https://github.com/tsujan/FeatherPad/issues/503#issuecomment-1236305302). Then everyone would be happy.
Thanks.

**Reply by tsujan** [reply]
FeatherPad opens multiple files in separate threads by design. Its side-pane mode provides not only a clean sorting but also an easy filtering. If you insist on avoiding view-items, you need a tabbed text editor that opens multiple files sequentially — most text editors do.

**Comment by cepamoi** [comment]
Sorry to insist, but I still don't understand why you completely reject a feature which has been requested by several users and effectively has some practical value for these users (but maybe not for you).
The implementation is probably not that difficult, and the risk to break something is low especially if the feature is made optional.
I can understand that this feature is not a priority to you and that you have more important things to do. Therefore I'm volunteering to implement it.
What do you think?
Thanks.
